feat: Add angle, colour, and multiline fields.

NeilFraser opened the pull request:
In preparation for removing said fields from Blockly core.

Commit:
feat: Three new fields

Angle, colour, multiline.

No Mocha tests.

Commit:
Add (most) tests

Some tests still commented out.

Commit:
Add (most) tests

Some tests still commented out.

Commit:
Activate angle validator tests

rachel-fenichel commented on plugins/field-angle/src/field_angle.ts:
```diff
@@ -0,0 +1,553 @@
+    }
+
+    // Allow individual settings to override the mode setting.
+    if (config.clockwise) this.clockwise = config.clockwise;
```
I think this fails if mode was `compass` but `config.clockwise` was false.

rachel-fenichel replied:
(Which would be a weird case.)

NeilFraser replied:
Fixed.

NeilFraser replied:
Lol.  Reverted.  The tests fail if this behaviour is altered.
Will add it to the proposed updates later.

rachel-fenichel commented on plugins/field-angle/src/field_angle.ts:
```diff
@@ -0,0 +1,553 @@
+  override initView() {
+    super.initView();
+    // Add the degree symbol to the left of the number,
+    // even in RTL (issue #2380).
```
Change to a full link since it's going into a new repository: https://github.com/google/blockly/issues/2380

NeilFraser replied:
Fixed.

rachel-fenichel commented on plugins/field-angle/src/field_angle.ts:
```diff
@@ -0,0 +1,553 @@
+    if (multiplier) {
+      const value = this.getValue() as number;
+      this.displayMouseOrKeyboardValue(value + multiplier * this.round);
+      e.preventDefault();
```
Why only call `preventDefault()` and `stopPropagation()` when there's a multiplier?

Edit: ah, I see, because otherwise it's another key and you ignore it. How about putting a default case in the `switch` that returns early, and then doing everything in the `if (multiplier)` section without the conditional?

NeilFraser replied:
There's a school of thought that each function should only have a single return statement (to the point where zealots disapprove of guard clauses).  I don't subscribe to this philosophy religiously, but it does have a point in often making program flow easier to understand.  With a single return one can add code anywhere in a function and know it will execute.  Whereas with multiple returns one has to read the whole function to know where one can add code.

While I think there are many cases where multiple returns makes code easier to read and maintain, the general case benefits from a single return.
http://wiki.c2.com/?SingleFunctionExitPointPlease

rachel-fenichel commented on plugins/field-angle/test/field_angle_test.mocha.js:
```diff
@@ -0,0 +1,381 @@
+const {FieldAngle} = require('../src/index');
+const {assert} = require('chai');
+
+const {
```
FYI I filed https://github.com/google/blockly-samples/issues/1676 about the type issues with the test helpers.

rachel-fenichel reviewed (commented):
First pass: I reviewed the angle files and left comments.

rachel-fenichel commented on plugins/field-multilineinput/src/field_multilineinput.ts:
```diff
@@ -0,0 +1,521 @@
+   * @param fieldElement The element to populate with info about the field's
+   *     state.
+   * @returns The element containing info about the field's state.
+   * @internal
```
Remove `@internal` tag, here and below, since these will be called by code outside of this plugin.

NeilFraser replied:
Done.

Commit:
Restore full testing of colour field.

Also address review comments so far.

Commit:
Generator tests don't belong in field tests.

This suite was testing generators, not the field.

Commit:
Reverting fix due to tested behaviour

Commit:
Add TypeScript dev dependencies.

...on all projects that use TS.

Commit:
Add TypeScript dev dependencies.

...on all projects that use TS.

NeilFraser committed:
Merge branch 'master' into fraser-triple

BeksOmega committed:
feat: add test block with resizable block in mutator (#1638)

* chore: clean up mutator test blocks

* chore: dedupe and reorganize code

* feat: add block with resizable block in mutator

mrkprdo committed:
fix: replace scrollToVisible_ with standard centerOnBlock (#1678)

* Fix: scrollToVisible to use ws.centerToBlock

* Fix: change to str compare to check function exist

* Fix: replace scrollToVisible with ws.centerOnBlock

github-actions[bot] committed:
Publish

 - @blockly/block-dynamic-connection@0.2.15
 - @blockly/block-plus-minus@4.0.14
 - @blockly/block-shareable-procedures@1.2.8
 - @blockly/block-test@3.1.0
 - @blockly/workspace-content-highlight@2.0.14
 - @blockly/continuous-toolbox@3.0.14
 - @blockly/plugin-cross-tab-copy-paste@2.0.17
 - @blockly/dev-tools@5.2.6
 - @blockly/disable-top-blocks@0.2.14
 - @blockly/field-bitmap@2.0.16
 - @blockly/field-colour-hsv-sliders@2.1.2
 - @blockly/field-date@6.0.14
 - @blockly/field-dependent-dropdown@1.1.2
 - @blockly/field-grid-dropdown@2.0.15
 - @blockly/field-slider@4.0.15
 - @blockly/fixed-edges@2.0.14
 - @blockly/keyboard-navigation@0.3.15
 - @blockly/plugin-modal@4.0.15
 - @blockly/renderer-inline-row-separators@0.2.11
 - @blockly/plugin-scroll-options@3.0.15
 - @blockly/plugin-serialize-disabled-interactions@3.0.2
 - @blockly/shadow-block-converter@2.0.14
 - @blockly/plugin-strict-connection-checker@2.0.14
 - @blockly/suggested-blocks@2.0.15
 - @blockly/plugin-typed-variable-modal@5.0.16
 - @blockly/workspace-backpack@3.1.3
 - @blockly/plugin-workspace-search@6.0.14
 - @blockly/zoom-to-fit@3.0.14

rachel-fenichel committed:
cleanup(examples): delete parcel example (#1681)

cpcallen commented on plugins/dev-scripts/package.json:
```diff
@@ -43,7 +43,7 @@
   },
   "peerDependencies": {
-    "typescript": "^4.3.2"
+    "typescript": "^5.0.2"
```
This should be replaced by PR #1684.

cpcallen commented on plugins/field-angle/package.json:
```diff
@@ -0,0 +1,64 @@
+    "blockly": "^9.3.1",
+    "chai": "^4.2.0",
+    "sinon": "^9.0.1",
+    "typescript": "^4.3.2"
```
You probably want to `npm install typescript@latest` here.

cpcallen committed:
chore(deps): Allow typescript v5 as peer dep of dev-scripts (#1684)

This should allow plugins to pull in a sufficiently recent version
of typescript that eslint will allow them to use the override
keyword.

cpcallen reviewed (changes requested):
Once the `typescript` dependencies are updated per comments below, run `rm -rf plugins/*/node_modules` then `npm ci` (both from repository root, not anywhere in `plugins/`), and then check in any changes to any `package-lock.json` file (which should _mostly_ be deletions of references to other packages in `plugins/` and their dependencies).

gonfunko committed:
fix: Clean the dist directory before every build. (#1685)

Commit:
Revert "Add TypeScript dev dependencies."

This reverts commit b580594e7c6bb543ff2e102dd2226854402b570b.

# Conflicts:
#	plugins/field-colour-hsv-sliders/package.json

Commit:
Revert "Add TypeScript dev dependencies."

This reverts commit b580594e7c6bb543ff2e102dd2226854402b570b.
